File: MATLAB/mass.m
function M = mass(nsd,ned,nn,coords,nel,nen,connect,materialprops)
% returns the global mass matrix
M = zeros(nn*ned,nn*ned);
elecoord = zeros(nsd,nen);
%%
% loop over elements
for ele = 1:nel
    % Extract coords of nodes, DOF for the current element
    for a = 1:nen
        elecoord(:,a) = coords(:,connect(a,ele));
    end
    %% compute the element mass matrix
    %
    %
    m = zeros(nen*ned,nen*ned);
    rho = materialprops(5);
    % set up quadrature points and weights
    npt = no_quad_pts(nsd,nen,0);
    xilist = quad_pts(nsd,nen,npt);
    weights = quad_weights(nsd,nen,npt);
    % loop over quadrature points
    for intpt = 1:npt
        xi = xilist(:,intpt);
        dNdxi = sfder(nen,nsd,xi);
        N = sf(nen,nsd,xi);
        % set up the jacobian matrix
        dxdxi = zeros(nsd,nsd);
        dxdxi = elecoord*dNdxi;
        dt = det(dxdxi);
        % Compute the element mass, no loop of j because M = delta_ij*M^~       
        for a = 1:nen
            for b = 1:nen
                for i = 1:ned
                    row = ned*(a-1)+i;
                    col = ned*(b-1)+i;
                    m(row,col) = m(row,col) + N(a)*N(b)*rho*dt*weights(intpt);  
                end
            end
        end
    end  
    %% scatter the element mass matrix into the global mass matrix
    %
    %
    for a = 1:nen
      for i = 1:ned
        for b = 1:nen
          for k = 1:ned
            row = ned*(connect(a,ele)-1)+i;
            col = ned*(connect(b,ele)-1)+k;
            M(row,col) = M(row,col) + m(ned*(a-1)+i,ned*(b-1)+k);
          end
        end
      end
    end
end
end

File: fortran/mass.f90
module mass
    implicit none
contains
    subroutine mass_matrix(mass)
        use read_file
        use shapefunction
        use integration
        
        real(8), dimension(no_nonzeros), intent(inout) :: mass ! The actual length should be less thant no_nonzeros
        real(8), dimension(nen*nsd,nen*nsd) :: mele
        real(8), dimension(nsd,nen) :: elecoord
        real(8), dimension(nsd,nen) :: eledof
        real(8), dimension(nsd,nsd) :: dxdxi
        real(8), allocatable, dimension(:,:) :: xilist
        real(8), allocatable, dimension(:) :: weights
        real(8), dimension(nen) :: N
        integer :: ele,a,b,i,j,npt,k,intpt,row,col
        real(8) :: det, rho
        real(8), dimension(nsd) :: xi
        real(8), dimension(nen,nsd) :: dNdxi 
        
        rho = materialprops(1)
        
        ! allocate
        npt = int_number(nsd,nen,0)
        allocate(xilist(nsd,npt))
        allocate(weights(npt))
        xilist = int_points(nsd,nen,npt)
        weights = int_weights(nsd,nen,npt)
            
        ! initialize
        mass= 0.
        
        ! loop over elements
        do ele=1,nel
            ! extract coords of nodes, and dofs for the element
            do a=1,nen
                elecoord(:,a) = coords(:,connect(a,ele))
            end do
            ! initialize
            mele = 0.
            ! loop over integration points
            do intpt=1, npt
                xi = xilist(:,intpt)
                dNdxi = sfder(nen,nsd,xi)
                N = sf(nen,nsd,xi)
                ! set up the jacobian matrix
                dxdxi = matmul(elecoord,dNdxi)
                ! determinant
                if (nsd == 2) then
                    det = dxdxi(1,1)*dxdxi(2,2) - dxdxi(1,2)*dxdxi(2,1)
                else if (nsd == 3) then
                    det = dxdxi(1,1)*dxdxi(2,2)*dxdxi(3,3) - dxdxi(1,1)*dxdxi(3,2)*dxdxi(2,3) &
                            - dxdxi(1,2)*dxdxi(2,1)*dxdxi(3,3) + dxdxi(1,2)*dxdxi(2,3)*dxdxi(3,1) &
                            + dxdxi(1,3)*dxdxi(2,1)*dxdxi(3,2) - dxdxi(1,3)*dxdxi(2,2)*dxdxi(3,1)
                end if
                ! compute the element mass matrix
                do a = 1, nen
                    do b = 1, nen
                        do i = 1, nsd
                            row = nsd*(a-1) + i
                            col = nsd*(b-1) + i
                            mele(row,col) = mele(row,col) + N(a)*N(b)*rho*det*weights(intpt)
                        end do
                    end do
                end do
            end do
            ! scatter
            do a=1,nen
                do i=1,nsd
                    row = nsd*(connect(a,ele)-1)+i
                    do b=1,nen
                        do k=1,nsd
                            col = nsd*(connect(b,ele)-1)+k
                            call addValueSymmetric(mass, row, col, mele(nsd*(a-1)+i,nsd*(b-1)+k))
                        end do
                    end do
                end do
            end do
        end do
        
        deallocate(xilist)
        deallocate(weights)
        
    end subroutine mass_matrix
end module
